packages/flutter/lib/src/widgets/undo_history.dart

```diff
@@ -0,0 +1,466 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+
+import 'actions.dart';
+import 'focus_manager.dart';
+import 'framework.dart';
+import 'text_editing_intents.dart';
+
+/// Provides undo/redo capabilities for a [ValueNotifier].
+///
+/// Listens to [value] and saves relevant values for undoing/redoing. The
+/// cadence at which values are saved is a best approximation of the native
+/// behaviors of a number of hardware keyboard on Flutter's desktop
+/// platforms, as there are subtle differences between each of the platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and calls [onTriggered] when a
+/// shortcut is triggered that would affect the state of the [value].
+///
+/// The [child] must manage focus on the [focusNode]. For example, using a
+/// [TextField] or [Focus] widget.
+class UndoHistory<T> extends StatefulWidget {
+  /// Creates an instance of [UndoHistory].
+  const UndoHistory({
+    super.key,
+    this.shouldChangeUndoStack,
+    required this.value,
+    required this.onTriggered,
+    required this.focusNode,
+    this.controller,
+    required this.child,
+  });
+
+  /// The value to track over time.
+  final ValueNotifier<T> value;
+
+  /// Called when checking whether a value change should be pushed onto
+  /// the undo stack.
+  final bool Function(T? oldValue, T newValue)? shouldChangeUndoStack;
+
+  /// Called when an undo or redo causes a state change.
+  ///
+  /// If the state would still be the same before and after the undo/redo, this
+  /// will not be called. For example, receiving a redo when there is nothing
+  /// to redo will not call this method.
+  final void Function(T value) onTriggered;
```

Comment:
If I understand how InputFormatters work correctly, I think this should not be an issue as they run before the value gets here. For example, I tried the CreditCardFormatter from the pattern_formatter package...if I type `1111` in rapid succession so that all ends up on the stack as one undo item, then press `2`, the value pushed onto the stack is `1111 2` (with the space added already by the formatter). When I hit undo, it correctly goes back to `1111` (no space). Pressing redo correctly changes the value to `1111 2`.

Reply:
No I meant what if the the undo value gets formatted for some reason.
Take your example , `1111 2` if you hit undo this method will be called with `1111`, what if the text format suddenly decide to format the incoming value `1111` to something else like `3333`.  If i understand the code correctly, this will create another undo history right after the undo. Should we throw in this case?

Reply:
Oh I see what you are saying...I did a RandomNumberFormatter to replace numbers with random digits...it does cause some issues (really what it messes up mostly is being able to redo). The current implementation seems to also have issues with this, so I don't think we are adding new issues here, at least.

So, do you or @justinmc have thoughts on whether that is more of a client concern or something the `UndoHistory` widget should handle?

Essentially, I think we could either handle it by preventing pushes while `onTriggered` is running or passing an optional parameter to `shouldChangeUndoStack` to let clients decide.

I'm leaning toward the latter, since it seems like this is a fairly specific case to EditableText and InputFormatters.

Reply:
I think this may be a corner case that is not so common in real life. I vote for just assert the value should stay the same after this method is called. If this does crash some use case, we can figure out how to handle it later.  Also at that point we may have a better picture on how we should support this use case

Reply:
Oh gotcha – I actually already implemented that with this commit: https://github.com/flutter/flutter/pull/98294/commits/b74a5db9dece2a9e40b511fe859844fd2436bce7

I think it works pretty well, though we could change that to make some kind of assert if `_push()` is called while `_duringTrigger` is true or something along those lines.

Reply:
I am not a big fan of duringTrigger API, It feel a bit confusing to the user if they don't know about this undo issue. I think we should hardcode the behavior for now.

Reply:
Gotcha. So...should we add an actual `assert` or just a guard clause to prevent pushing that new value onto the undo stack?

I've been playing with assertions in a couple of cases and run into some issues. If I put the assert in `_update`, I end up with failing tests because the value doesn't get updated to the new value on android when composing changes. If I put the assert in `_push`, I can't seem to properly catch that assertion failure in tests because that is called as the result of a change notifier, and I can't figure out how to catch that properly.

At any rate – here's what I ended up going with for now: https://github.com/flutter/flutter/pull/98294/commits/a26bd352845554ce7702380c228b95cf45e4778b – hard coding that behavior in `_push` and updating the documentation for `onTriggered` to describe the behavior.

Reply:
If we prevent pushing the new value on to the stack, wouldn't that makes the current value on the top of the stack to be inconsistent with whats in the textfield?

Can we add an assert in line 148 in flutter/lib/src/widgets/undo_history.dart
right after the widget.onTriggered(nextValue); is called to verify the value doesn't change?

Reply:
So, the issue I run into if I try to do that is that composing acts a little weird – I end up in a situation where the following tests fail:
TextEditingHistory does save composing changes on Android (variant: TargetPlatform.android)
TextEditingHistory saves right up to composing change even when throttled (variant: TargetPlatform.android)

I think the issue is that the TextEditingValue ends up changing in that case even though the actual text part of it doesn't. I tried checking that the value is either the new value or the last value, but that didn't resolve the issue.

Reply:
Do you have a feature branch for the code that causes the error so that i can test and see what was wrong?

Reply:
Sure – https://github.com/twinsunllc/flutter/tree/feature/undo-redo-gestures-with-assertion

Reply:
@chunhtai Have you had a chance to take a look at this? Just want to put it on your radar again after the holidays.

Reply:
sorry for the delay, I will take a look today

Reply:
The issue is the editable text does not preserve composing in the onTrigger callback. Is this a intended behavior? also cc @justinmc 

If so, the assert should just check for text and selection and omit composing.

Also it should only check for nextValue, it doesn't make sense to check for previousValue. Unless there is a use case I was missing?

Reply:
Changes to composing text are purposely not included in the undo stack (except on Android because Gboard on Android uses composing for everyday English words).  See https://github.com/flutter/flutter/pull/108765.

Reply:
It's been a little bit since I looked at this in detail, so I will have to refresh my own memory when I get a chance, but if I recall correctly, the reason I was checking the `lastValue` as well is the composing issue – though it still fails those tests, so I'm not sure that works anyway.

Part of the issue is that this is not specific to text editing, so we don't have access to the underlying elements of a TextEditingValue here, we would need to either create an interface to expose some sort of undo/redo-specific value we want compared or farm it out to the clients who are using the UndoHistory. That was my original thinking with adding the optional parameter to shouldChangeUndoStack, that way the clients get to decide how to compare those values and whether changes during the onTrigger are a problem or not.

Reply:
Is your branch [undo-redo-gestures-with-assertion](https://github.com/twinsunllc/flutter/tree/feature/undo-redo-gestures-with-assertion) still up to date, and do you have example code that causes the bugs mentioned here?

I checked out that branch and saw the failures in editable_text_test.dart.  Your change to drop the composing range in onTriggered [0] might not be correct.  Trying it on a native Android app, the composing region seems to remain when undoing a partial word (or maybe it's just reopening).  What if we kept the composing range in editable_text.dart's onTriggered like it was originally, and then modified the test to expect the composing range to exist?

Or maybe I'm oversimplifying and I need to see some example code for the bugs I'm forgetting about.

Here's the original code that didn't drop the composing range:

https://github.com/flutter/flutter/blob/234452fef20f4b1ca199b4a177ed3222979aa2d7/packages/flutter/lib/src/widgets/editable_text.dart#L4285-L4286

And here is a failing test that confirms that the composing range is gone, but maybe it should be changed to include the original composing range:

https://github.com/flutter/flutter/blob/234452fef20f4b1ca199b4a177ed3222979aa2d7/packages/flutter/test/widgets/editable_text_test.dart#L12735

Reply:
I will trust your judgement on the right way to treat the composing region, I don't know much about it. Tracking the composing changes seems to work, so I went ahead and made those changes and brought that and the assertion over to this branch.

Here's an example that trips the assertion (I modified one of the formatters from the `pattern_formatter` package as a starting point):
<details>
<summary>Example assertion error</summary>

```
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:pattern_formatter/pattern_formatter.dart';

final DateTime launchTime = DateTime.now();

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'Flutter Code Sample';

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextEditingController _controller = TextEditingController();
  final FocusNode _focusNode = FocusNode();
  final UndoHistoryController _undoController = UndoHistoryController();

  final FocusNode _counterFocus = FocusNode();
  final ValueNotifier<int> _counter = ValueNotifier<int>(0);
  final UndoHistoryController _counterUndoController = UndoHistoryController();

  TextStyle? get enabledStyle => Theme.of(context).textTheme.bodyMedium;
  TextStyle? get disabledStyle => Theme.of(context).textTheme.bodyMedium?.copyWith(color: Colors.grey);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(launchTime.toIso8601String()),
            UndoHistory<int>(
              value: _counter,
              onTriggered: (int newValue) {
                _counter.value = newValue;
                setState(() {});
              },
              focusNode: _counterFocus,
              controller: _counterUndoController,
              child: Focus(
                focusNode: _counterFocus,
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Text('Current value: ${_counter.value}'),
                    Text('Has Focus: ${_counterFocus.hasFocus}'),
                    TextButton(
                      child: const Text('Increment'),
                      onPressed: () {
                        _counter.value = _counter.value + 1;
                        setState(() {});
                        _counterFocus.requestFocus();
                      },
                    ),
                  ],
                ),
              ),
            ),
            ValueListenableBuilder<UndoHistoryValue>(
              valueListenable: _counterUndoController,
              builder: (context, value, _) {
                return Row(
                  children: [
                    TextButton(
                        child: Text('Undo', style: value.canUndo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _counterUndoController.undo();
                        }),
                    TextButton(
                        child: Text('Redo', style: value.canRedo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _counterUndoController.redo();
                        }),
                  ],
                );
              },
            ),
            TextField(
              maxLines: 4,
              controller: _controller,
              focusNode: _focusNode,
              undoController: _undoController,
              inputFormatters: [
                RandomNumberFormatter(),
              ],
            ),
            ValueListenableBuilder<UndoHistoryValue>(
              valueListenable: _undoController,
              builder: (context, value, _) {
                return Row(
                  children: [
                    TextButton(
                        child: Text('Undo', style: value.canUndo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _undoController.undo();
                        }),
                    TextButton(
                        child: Text('Redo', style: value.canRedo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _undoController.redo();
                        }),
                  ],
                );
              },
            ),
          ],
        ),
      ),
    );
  }
}

class RandomNumberFormatter extends NumberInputFormatter {
  static final RegExp _digitOnlyRegex = RegExp(r'\d+');
  static final FilteringTextInputFormatter _digitOnlyFormatter = FilteringTextInputFormatter.allow(_digitOnlyRegex);

  RandomNumberFormatter();

  @override
  String _formatPattern(String digits) {
    StringBuffer buffer = StringBuffer();

    for (int i = 0; i < digits.length; i++) {
      buffer.write(math.Random().nextInt(9));
    }
    return buffer.toString();
  }

  @override
  TextEditingValue _formatValue(TextEditingValue oldValue, TextEditingValue newValue) {
    return _digitOnlyFormatter.formatEditUpdate(oldValue, newValue);
  }

  @override
  bool _isUserInput(String s) {
    return _digitOnlyRegex.firstMatch(s) != null;
  }
}

///
/// An abstract class extends from [TextInputFormatter] and does numeric filter.
/// It has an abstract method `_format()` that lets its children override it to
/// format input displayed on [TextField]
///
abstract class NumberInputFormatter extends TextInputFormatter {
  TextEditingValue? _lastNewValue;

  @override
  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
    /// nothing changes, nothing to do
    if (newValue.text == _lastNewValue?.text) {
      return newValue;
    }
    _lastNewValue = newValue;

    /// remove all invalid characters
    newValue = _formatValue(oldValue, newValue);

    /// current selection
    int selectionIndex = newValue.selection.end;

    /// format original string, this step would add some separator
    /// characters to original string
    final newText = _formatPattern(newValue.text);

    /// count number of inserted character in new string
    int insertCount = 0;

    /// count number of original input character in new string
    int inputCount = 0;
    for (int i = 0; i < newText.length && inputCount < selectionIndex; i++) {
      final character = newText[i];
      if (_isUserInput(character)) {
        inputCount++;
      } else {
        insertCount++;
      }
    }

    /// adjust selection according to number of inserted characters staying before
    /// selection
    selectionIndex += insertCount;
    selectionIndex = math.min(selectionIndex, newText.length);

    /// if selection is right after an inserted character, it should be moved
    /// backward, this adjustment prevents an issue that user cannot delete
    /// characters when cursor stands right after inserted characters
    if (selectionIndex - 1 >= 0 && selectionIndex - 1 < newText.length && !_isUserInput(newText[selectionIndex - 1])) {
      selectionIndex--;
    }

    return newValue.copyWith(text: newText, selection: TextSelection.collapsed(offset: selectionIndex), composing: TextRange.empty);
  }

  /// check character from user input or being inserted by pattern formatter
  bool _isUserInput(String s);

  /// format user input with pattern formatter
  String _formatPattern(String digits);

  /// validate user input
  TextEditingValue _formatValue(TextEditingValue oldValue, TextEditingValue newValue);
}
```
</details>

If I swap in `CreditCardFormatter` for the `RandomNumberFormatter`, the assertions don't get thrown, so it seems like this is a reasonable solution.

Reply:
I tried this out again and it seems to work for me on both iOS and Android with Gboard, where it correctly includes composing state in the undo history.  I also tried the example above and got the assertion.  Thanks!